### PR TITLE
Add WAF as a top level resource

### PR DESCRIPTION
**Pull request by azhao155**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [x] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by azhao155** [comment]
This pr is the duplicate of the following one:
https://github.com/Azure/azure-rest-api-specs/pull/5185

the resource api version is 12/01/2018, so it should be fine to merge into February branch. 

@number213 , please comment, since you are the reviewer for the above pr

**Review by akshaysngupta — APPROVED** [review]

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json
Comment (number213) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: `Conditon`
Typo? Examples don't contain this property, could you please verify how it's spelled on backend?

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json
Comment (number213) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: This enum has the same name it's parent definition has, this could cause collisions in generated SDK. Plus as the name of enum it's not really clear. Please rename it to something like `WebApplicationFirewallPolicyResourceState`

**Review by anton-evseev — CHANGES_REQUESTED** [review]
LGTM except a few nitpicks.
Also, regarding definitions `PolicySettings`, `CustomRule`, `Transform`, `MatchVariable` and `MatchCondition`. All Network models in SDK exist in one namespace, so having a model like `CustomRule` without context may be confusing. I would prefix all of them with `WebApplicationFirewall` (e.g. `WebApplicationFirewallCustomRule`)
@akshaysngupta @dsgouda what do you think?

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json
Comment (number213) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: Please add `security` property to fix CI errors
See https://github.com/Azure/azure-rest-api-specs/blob/master/specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/endpointService.json#L18 for reference

**Review by anton-evseev — CHANGES_REQUESTED** [review]

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/examples/WafPolicyGet.json
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/examples/WafPolicyGet.json]: nit: Please use 2 spaces instead of 4 for formatting

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: Please add a default error response

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: Please mark this as readonly

**Review by dsgouda — CHANGES_REQUESTED** [review]

**Comment by MikhailTryakhov** [comment]
@azhao155  please escalate fixes here, release is on Monday and you're close to miss it

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json
Comment (azhao155) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: I will change definition of custom rule into WebApplicationFirewallCustomRule

Comment (azhao155) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: we can't make the change now, since NRP already has the propety named custom rule, and property name in the swagger and nrp should match

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json
Comment (azhao155) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: Yes, that's what we specify in the backend, so it has to be match

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json
Comment (azhao155) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: made the change

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json
Comment (azhao155) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: added

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json
Comment (azhao155) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: changed

File: specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json
Comment (azhao155) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json]: I did the powershell testing, if add default error response, it will generate different exception which won't be caught. Please reference the applicationGateway.json delete operation in the network.

**Comment by azhao155** [comment]
@MikhailTryakhov @dsgouda @number213 I have addressed comments, please check, it's critical for us and we need to catch the February release.

**Review by dsgouda — APPROVED** [review]
LGTM will wait for @number213 's approval

**Comment by azhao155** [comment]
@asridharan this is needed to be into febuary release

**Comment by azhao155** [comment]
@MikhailTryakhov @dsgouda @number213  i saw it has out of memory issue, could you help us to address it? it's urgent for us.

**Comment by dsgouda** [comment]
@jhendrixMSFT @sergey-shandar could you take a look

**Comment by azhao155** [comment]
@jhendrixMSFT @sergey-shandar Thanks for the fix, but it's still failing, we need to merge this pr by EOD, please help.

**Comment by dsgouda** [comment]
@azhao155 Some of these are acceptable failures, will merge the PR once all the builds are done.

**Comment by azhao155** [comment]
Thanks for @dsgouda explanation. Thanks for the help again!

**Pull request merged**
